### ISSUE TITLE
Support using -MD instead of -MMD

### ISSUE THREAD

**Issue by haasn** [mesonbuild/meson#2484]
From gcc documentation:

```
       -MMD
           Like -MD except mention only user header files, not system header files.
```

Sometimes this is not what the developer wants. Allowing it to be tunable or overridable somehow would be great. It seems this is a hard-coded meson behavior:

https://github.com/mesonbuild/meson/blob/33c2bbc6e889ed912942b2e3507deeafc691c6a2/mesonbuild/compilers/c.py#L94

As a motivating case, the `mpv` project needs to rebuild a number of files whenever the FFmpeg headers change, because FFmpeg's ABI policy is... unorthodox; and header changes can randomly break everything unless the corresponding files are rebuilt.

**Comment by jpakkane** [comment]
Could we just make it the default? It is more reliable but OTOH it slows things down due to extra lookups. On Linux this is probably fine but what about Windows?

**Comment by haasn** [comment]
How much slower is it really? Is there a simple wya to override this somehow for testing? (I tried CFLAGS but it seems to have gotten ignored)

**Comment by jpakkane** [comment]
> Is there a simple wya to override this somehow for testing?

Only by editing the source code of Meson.

**Comment by nirbheek** [comment]
It would be interesting to test building [gst-build](http://cgit.freedesktop.org/gstreamer/gst-build/) with both those and compare the difference. Might be enough to generate a build.ninja and then do a sed on it.

**Comment by haasn** [comment]
Oh, yes, editing the resulting build.ninja is certainly the easiest way of testing it - the only two lines to change were right at the top for me. As expected, the “no work to do” invocation takes exactly 5ms both before and after on my small test project. I'll try some larger projects and report back.

**Comment by haasn** [comment]
Tested it on gst-build; after replacing 'MMD' by 'MD' in the build.ninja the 'no work to do' invocation requires 120ms instead of 113ms on average; i.e the extra time spent checking system-wide headers.

To double check it was working, I touched some random headers in /usr/include - all the affected files were successfully rebuilt. The time for a fresh build (after `ninja clean`) went from 38.63s to 38.65s.

In conclusion, I think using `-MD` instead of `-MMD` would be a safe default, performance-wise.

**Comment by nirbheek** [comment]
I think C++ projects are affected by this more than C ones, and Windows is worse because `windows.h` includes a *lot* of things. But if the performance difference is as small as you measured for those too, we should definitely change this.

**Comment by nirbheek** [comment]
@haasn would you like to work on this? The change is relatively straightforward.

**Comment by haasn** [comment]
I guess the open questions are:

- Do we use `-MD` by default always, or do we make it configurable?
- Does it cause performance issues for any windows C++ projects?

**Comment by nirbheek** [comment]
Large windows C++ projects will probably use MSVC anyway, won't they? I say we do it by default and see if anyone complains. Adding configuration options prematurely just increases maintenance burden.

**Comment by haasn** [comment]
This was addressed in 8eaa0a273. Closing this issue.
